### PR TITLE
Fixes #2400, ServletContainerInitializer support

### LINKED ISSUE #2400: ServletContainerInitializer implementations are never invoked
I want to use [Sentry](https://sentry.io/) in my Quarkus JVM App. For that, they provide the following dependency that contains a `io.sentry.servlet.SentryServletContainerInitializer`:

```xml
    <dependency>
      <groupId>io.sentry</groupId>
      <artifactId>sentry</artifactId>
      <version>1.7.22</version>
    </dependency>
```

However the initializer is never triggered when the dependency is added to a Quarkus app.

### PR DESCRIPTION
**Pull request by stuartwdouglas**

### CODE REVIEW

File: extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java
Comment (gsmet) [inline extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java]: I think you should use `ServiceUtil` here.

You could have comments in the file and `ServiceUtil` manages all of that (and is used everywhere else).

File: extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java
Comment (gsmet) [inline extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java]: Can't say I'm a big fan of you using single letter variables, every time you have a for loop. It's quite hard to know what they are referring to. (This is a general comment for the whole patch)

It's really just cosmetic but it's not the first time I have a hard time following this sort of patterns in the code so I thought I might as well raise the issue.

**Review by gsmet — CHANGES_REQUESTED** [review]
Added 2 comments, one purely cosmetic, one not so much.

File: extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java
Comment (stuartwdouglas) [reply extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java]: I was not aware of ServiceUtil, however I am not sure that its current usage is safe.

In particular the use of getClass().getClassLoader() will return a class loader that has all the deployment related artifacts on the class path. As long as none of these contain the services this is fine, however if a deployment artifact or its dependency contains this service then we will have problems. ServletContainerInitializer's are fairly widely used, so it is possible this will cause problems down the road. 

This needs a bit of thought as to how to handle it properly.

Comment (gsmet) [reply extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java]: Yes, for now, the classloader situation is far from being perfect. I know @dmlloyd has a plan for that.

Not totally sure if we should try to find a solution now or just collect the issues we currently have to nourish this reflection.

Comment (dmlloyd) [reply extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java]: I don't think getting services from the deployment class path is a good idea.  We talked about the possibility having a synthetic class loader which represents the application but that would not be for situations like this.  Using `ServiceUtil` would not be right for this situation.

Comment (stuartwdouglas) [reply extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java]: Why would a synthetic ClassLoader not be right for this case? You could just pass it into ServiceUtil and it should limit services to the application

Comment (dmlloyd) [reply extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java]: You don't *need* a `ClassLoader` for this though, because it is a build step so you could scan the various application roots instead.

That said, if we were to introduce a new class loader for this (I was thinking we want to expose the three or so of them by build items) then there are a few other different useful applications for it.

**Comment by mkouba** [comment]
@dmlloyd @gsmet Pls approve this PR if all of your comments were addressed.

**Comment by dmlloyd** [comment]
I'll leave it to @gsmet, since I don't really have any opinion one way or another other than my comments about potential class loader stuff.

**Review by gsmet — APPROVED** [review]
Sorry, I was at a conference and now there's a conflict.

It looks good to me, once the conflict is fixed.

**Comment by mkouba** [comment]
@stuartwdouglas Pls rebase.

**Pull request merged**
